ocean/helium/snapshots.py

```diff
@@ -131,10 +136,267 @@ def download_file_from_snapshot(src, dst, snapshothash):
+            path: path to package
+
+        Returns:
+            A new package of that path
+
+        Raises:
+            when path doesn't exist
+        """
+        # TODO: anything but local paths
+        # TODO: deserialization metadata
+        data = {}
+        meta = {}
+        src_path = pathlib.Path(path)
+        files = src_path.rglob('*')
+        for f in files:
+            if not f.is_file():
+                continue
+            entry = {
+                'hash': hash_file(f),
+                'size': os.path.getsize(f),
+                'user_meta': {},
+                'physical_keys': [{
+                    'type': 'local_file',
+                    'path': os.path.abspath(f)
+                }]
+            }
+            data[str(f)] = entry
+        return Package(data, meta)
+
+    def get(self, logical_key):
+        """
+        Gets object from local_key and returns it as an in-memory object.
+
+        Args:
+            logical_key: logical key of the object to get
+
+        Returns:
+            A deserialized object from the logical_key or
+                a stream of bytes if deserialization info is missing
+
+        Raises:
+            KeyError: when logical_key is not present in the package
+            physical key failure
+            hash verification fail
+        """
+        entry = self._data[logical_key]
+        physical_keys = entry['physical_keys']
+        physical_key = physical_keys[0] # TODO: support multiple physical keys
+        stream = dereference_physical_key(physical_key)
+        # TODO: verify hash
+        if 'target' in entry:
+            # TODO: dispatch on target to deserialize
+            raise NotImplementedError
+
+        return stream
+
+    def get_file(self, logical_key, path):
+        """
+        Gets object from logical_key inside the package and saves it to path.
+
+        Args:
+            logical_key: logical key inside package to get
+            path: where to put the file
+
+        Returns:
+            None
+
+        Raises:
+            logical key not found
+            physical key failure
+            fail to create file
+            fail to finish write
+        """
+        raise NotImplementedError
+
+    def get_meta(self, logical_key):
+        """
+        Returns user metadata for specified logical key.
+        """
+        entry = self._data[logical_key]
+        return entry['user_meta']
+
+    def dump(self, path):
+        """
+        Serializes this package to a file at path.
+
+        Args:
+            path: where to serialize the package to
+
+        Returns:
+            None
+
+        Raises:
+            fail to create file
+            fail to finish write
+        """
+        with open(path, mode='w') as f:
+            with jsonlines.Writer(f) as writer:
+                writer.write(self._meta)
+                for logical_key, obj in self._data.items():
+                    obj['logical_key'] = logical_key
```

Comment:
You modifying `self._data` here... probably don't want that

Reply:
correct